### PR TITLE
AP_Frsky_Telem: added waypoints to FrSky PassThrough telemetry at 0x5009

### PR DESCRIPTION
**Pull request by yaapu**
This adds waypoint data to the frsky passthrough protocol.
All navigation information is updated from vehicle code with a callback that updates the _nav_info struct, the approach is similar to what the osd code does, it supports Copter, Plane and Rover.
The new 32bit telemetry packet holds

- wp number encoded as 10bits (1023 waypoints max)
- wp distance encoded as 12 bits (102.3Km max, 10bits + 2 for 10^power)
- wp xtrack error encoded as 6 bits (+/- 150m, 4 bits + 1 for 10^power +1 for sign)
- wp bearing offset from cog in 45° multiples encoded as 3 bits

The new packet is sent at 1Hz with DIY id 0x5009

The total number of mission commands is sent as parameter 6.

Code has been tested in SITL and flown on copter 3.6

### CODE REVIEW

File: APMrover2/APMrover2.cpp [hunk 1]
```diff
@@ -330,4 +333,19 @@ void Rover::update_current_mode(void)
     control_mode->update();
 }
 
+#if FRSKY_TELEM_ENABLED == ENABLED
+void Rover::update_nav_info()
+{
+    AP_Frsky_Telem::NavInfo nav_info;
+    //
```
Comment (peterbarker) [inline APMrover2/APMrover2.cpp]: These lines are weird.

File: ArduCopter/ArduCopter.cpp [hunk 2]
```diff
@@ -594,5 +597,18 @@ void Copter::publish_osd_info()
     osd.set_nav_info(nav_info);
 }
 #endif
-
+#if FRSKY_TELEM_ENABLED == ENABLED
+void Copter::update_nav_info()
+{
+    AP_Frsky_Telem::NavInfo nav_info;
+    //
+    nav_info.wp_distance = flightmode->wp_distance() * 1.0e-2f;
```
Comment (peterbarker) [inline ArduCopter/ArduCopter.cpp]: How are you determining the applicability of this data on the other end?

wp_distance means a lot of things in a lot of modes.... just not all the same.

File: ArduPlane/ArduPlane.cpp [hunk 3]
```diff
@@ -963,5 +966,18 @@ void Plane::publish_osd_info()
     osd.set_nav_info(nav_info);
 }
 #endif
-
+#if FRSKY_TELEM_ENABLED == ENABLED
+void Plane::update_nav_info()
+{
+    AP_Frsky_Telem::NavInfo nav_info;
+    //
+    nav_info.wp_distance = auto_state.wp_distance;
```
Comment (peterbarker) [inline ArduPlane/ArduPlane.cpp]: This isn't like all the others if I'm thinking right.  Some combination of guided_WP_loc and next_WP_loc would be more analogous

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp [hunk 4]
```diff
@@ -778,6 +786,44 @@ uint32_t AP_Frsky_Telem::calc_attiandrng(void)
...
+    // wp number, max is 1023 encoded as 10bits
+    waypoint = MIN(_nav_info.wp_number,WAYPOINT_NUMBER_LIMIT);
+    // max wp distance is 1023*100m = 102.3Km encoded as 12bits: 10 + 2 for 10^power
+    waypoint |= prep_number(MIN(_nav_info.wp_distance,WAYPOINT_DISTANCE_LIMIT),3,2) << WAYPOINT_DISTANCE_OFFSET;
+    // xtrack error, max is +/- 150m encoded as 6bits: 4 + 1 for 10^power + 1 for sign
+    waypoint |= prep_number(MIN(_nav_info.wp_xtrack_error,WAYPOINT_XTRACK_LIMIT),1,1) << WAYPOINT_XTRACK_OFFSET;
+    const bool gotGPS = (AP::gps().status() >= AP_GPS::GPS_OK_FIX_2D);
+    //
+    if (gotGPS) {
+        const int32_t cog = AP::gps().ground_course_cd(); //cog in centidegrees
```
Comment (peterbarker) [inline libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: Are you assuming you're flying forward here?

File: APMrover2/APMrover2.cpp (hunk 1 above)
Comment (yaapu) [reply APMrover2/APMrover2.cpp]: Hi Peter thanks for taking a look.
Could you please help me for I cannot see the "weirdness"!

File: ArduCopter/ArduCopter.cpp (hunk 2 above)
Comment (yaapu) [reply ArduCopter/ArduCopter.cpp]: Yes, wp_distance means a lot of different things, I noticed it in my tests in RTL for instance and I'm sure each mode has his own use of the waypoint "subsystem", this though is not as bad as it might seem from "the other end" point of view since the library already sends frametype and flight modes. In my tests I added some lua code to properly display the data in a proper context.
The basic idea is to send data as raw as possible and eventually process it (apply some logic) at the rendering side.

File: ArduPlane/ArduPlane.cpp (hunk 3 above)
Comment (yaapu) [reply ArduPlane/ArduPlane.cpp]: Hi, I'm still learning my way through vehicle code, I "borrowed" this from the recent OSD code 
https://github.com/ArduPilot/ardupilot/blob/c0af0cfe70a8a102b1d7dfa436a46d7c32039ce3/ArduPlane/ArduPlane.cpp#L967-L979
and I must admit I simply I assumed it was all right. having said that it really is OK with me to improve on the quality of what I proposed!

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp (hunk 4 above)
Comment (yaapu) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: ok this is another good question.
As I said I took inspiration from OSD code 
https://github.com/ArduPilot/ardupilot/blob/c0af0cfe70a8a102b1d7dfa436a46d7c32039ce3/libraries/AP_OSD/AP_OSD_Screen.cpp#L849
but I noticed it was based on yaw which in my opinion did not make much sense for vehicles able to move while rotating. I thought that cog was a better choice and this is the result.
I did not make any assumption but since this is all new to me I might have missed the obvious!

File: APMrover2/APMrover2.cpp (hunk 1 above)
Comment (OXINARF) [reply APMrover2/APMrover2.cpp]: I think @peterbarker is talking about the blank lines with just `//` in it.

**Comment by OXINARF** [comment]
@yaapu Peter already made some good comments and I'll leave the library part for the library maintainer (@floaledm), so I'll just ask that you change the commit messages to comply with our rules (http://ardupilot.org/dev/docs/submitting-patches-back-to-master.html).

**Comment by yaapu** [comment]
Thanks @OXINARF more than happy to comply,  but frankly I thought my commit comments where already all right.
Could you please help me to correct them or be more specific as to how to improve them, I wrote them as <subsystem> : brief description with a longer text on a separate paragraph.

**Comment by OXINARF** [comment]
@yaapu Interestingly they are indeed almost OK, but all of them have a `modified: [...]` list at beginning and ending of each message. You just need to remove that.
By the way, you also need to rebase on master as there is a conflict.

File: ArduCopter/ArduCopter.cpp (hunk 2 above)
Comment (floaledm) [reply ArduCopter/ArduCopter.cpp]: I might be missing a mode or two here but, upon a cursory reading, I find that wp_distance(), as it relates to its use in Copter only, gets us either:
- horizontal distance to destination in cm for RTL, SmartRTL, Auto, and Guided_WP modes
- horizontal distance to position target in cm for Loiter, Circle, and Guided_PosVel modes
That doesn't seem to represent a lot of different things. What am I missing?
By the way, Alessandro is right in that flightmode->wp_distance() is used by Copter::send_nav_controller_output() which is what is being sent to GCS via MAVLink.

File: ArduPlane/ArduPlane.cpp (hunk 3 above)
Comment (floaledm) [reply ArduPlane/ArduPlane.cpp]: Similarly, auto_state.wp_distance is what is used by Plane::send_nav_controller_output() to send nav info via MAVLink...

We may also need to consider the case of quadplanes in vtol: https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/GCS_Mavlink.cpp#L181

File: libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp (hunk 4 above)
Comment (floaledm) [reply libraries/AP_Frsky_Telem/AP_Frsky_Telem.cpp]: Again, I would look at the NAV_CONTROLLER_OUTPUT MAVLink message for inspiration (careful, as it is used differently depending on the vehicle, from what I understand of the ArduPilot code):
https://github.com/ArduPilot/mavlink/blob/master/message_definitions/v1.0/common.xml#L3245

**Comment by floaledm** [comment]
Hi all,

Sorry for not being able to participate in this discussion earlier. This is a nice effort to add waypoint information to the FrSky telemetry lib, with attention to minimizing bandwidth usage, so thanks Alessandro! I too would love to get access to waypoint info so I hope we can get this pushed through within a reasonable timeframe.

That being said, I think it would be a mistake to simply delve into the details of the proposed code changes without first discussing and agreeing on the best implementation for waypoint info, especially given the constraints of the FrSky link and also since the passthrough protocol is not modular (in it's current implementation, we can't configure what gets sent and what doesn't). Ideally, it would be nice to have two-way FrSky telemetry implemented (OpenTX already supports it) so that the passthrough protocol could become modular (and therefore better manage bandwidth!). It could then, among other things, feature a polling mechanism for retrieving waypoint data, similar to what is done by ground control stations via MAVLink. Short of a two-way link, I agree that we are stuck with periodically sending very selective info about the waypoints, just as we do for all other info in this lib.

So, in the context of a one-way FrSky downlink from the flight controller to a RC controller display (the most common usage), I'd like to first discuss what information we'd like to see about waypoints (some of which could be calculated from other info). From there, we could agree on which info is most crucial and also which resolution and offsets are acceptable for it, so that we can work our way towards getting the most relevant/useful info packetized in as few packets/bits as practicable. I understand Alessandro has already determined all this on his own, but I'd like the process to be collaborative in an attempt to accommodate the needs of as many users as possible. I don't know if Alessandro has looked into it (he mentioned using OSD code), but the vehicle specific info found in each of the NAV_CONTROLLER_OUTPUT MAVLink messages seems like a good place to start...
https://github.com/ArduPilot/ardupilot/blob/master/ArduCopter/GCS_Mavlink.cpp#L159
https://github.com/ArduPilot/ardupilot/blob/master/ArduPlane/GCS_Mavlink.cpp#L181
https://github.com/ArduPilot/ardupilot/blob/master/APMrover2/GCS_Mavlink.cpp#L104

Based on my review of the code in this PR, Alessandro would like to see (apologies in advance for any mistake - please correct if necessary),

Sent periodically every second:
- target waypoint number (aka current nav index)
- horizontal distance between current location and target waypoint
- horizontal crosstrack error (actual vs desired position along the track)
- waypoint bearing expressed as angle between ground course and wp direction, rounded to the nearest 45° increment (total of 8 possible "sectors," spanning 360°)

Sent once every few seconds (as a FrSky lib "parameter"):
- number of waypoints in the mission

I personally would like a better resolution on the waypoint bearing (e.g., 3°, as in the home bearing). I also think it would be interesting to get the altitude of the waypoint (relative to the vehicle or not, which could be compared to the vehicle altitude), and maybe the altitude error along the track if we already are looking at the crosstrack error, in which case, we may need two packets to fit everything... In terms of error, NAV_CONTROLLER_OUTPUT also features airspeed error but not sure if it's that useful. When the vehicle is more than a little off track, I haven't checked but would there be a message sent to alert the user, which could alleviate the need to send the crosstrack error values? We need to make sure we minimize (read "get rid of" here) redundant info...
Not sure how to achieve this (maybe derive from computations - it would be interesting to look at how it's done on ground stations) but I'd like to see an ETA to the waypoint as well (for loiter/circle, it would be nice also to know the time left to hold the pattern?). For me, waypoint number or total number of waypoints wouldn't be that useful since this is not for an interface where waypoints are editable, and the number of the waypoint currently tracked is already provided via text message ("command #").

So, I would personally like to see a waypoint packet periodically sending no more than this:
- horizontal and vertical distance between current location and target waypoint
- waypoint bearing expressed as angle between front of vehicle and wp direction, rounded to something like the nearest 3°
- ETA if it cannot be derived
I would also consider info that would give me a heads up on what the vehicle would do next, which could take the form of data about the next waypoint in the mission. Not sure on the details though...

But please participate and provide your thoughts. This is meant as an open discussion and we will come to the best solution together!

Cheers.

P.S.: I don't fly plane or drive rovers, so my experience on multiple vehicle types is limited to my understanding of the code...

Side note on scheduled tasks: It's definitely a valid approach. So much so that there used to be scheduled tasks for the FrSky lib in each vehicle, but we worked to clean up the code and remove these unnecessary tasks as there is no need to periodically and independently update things such as the nav_info. The nav_info needs to be made available to the FrSky lib (just like it is done for the rest of vehicle specific info), so that the FrSky lib updates its variables based on retrieving nav_info when it is most judicious, which seems to be right before that info needs to be packetized for transmission, instead of periodically as is done with scheduled tasks. This way, there is no unnecessary calculation cycles and the info is as fresh as it can be. Of course, if someone has a different perspective on this, I'd love to hear your thoughts! Maybe there's a better way altogether to get this done.

**Comment by floaledm** [comment]
I'd like to emphasize that this is really the time for anybody to pitch in... I don't know if we need to advertise inviting comments using other channels like the discuss forum. We need to decide at which point to end the comment period on what info should conceptually be included to make sure this PR gets through in a timely manner.

**Comment by magicrub** [comment]
perhaps @floaledm and @yaapu could join us on the dev call? I see that it had the tag, and it was removed, but maybe it needs more discussion? @OXINARF you removed the tag, what do you think?

http://ardupilot.org/dev/docs/ardupilot-mumble-server.html

**Comment by yaapu** [comment]
Hi @magicrub the PR has been briefly discussed in the devcall but the focus has been on how to best implement it and not on what to send, I guess the time to decide what to send is now as Florent noted.
Personally I'd also like to have the current waypoint number and total number of waypoints along the info florent proposes.
If we agree on using 2 packets like 0x5009 and 0x500A there's room for higher reslution.

To recap :
- 10 bits, total number of waypoints (max 1023) 
- 10 bits, current waypoint  (max 1023)
- 12 bits horizontal distance max 102.3Km
- ?? bits vertical distance between current location and target waypoint
- 7 bits, waypoint bearing with 3° resolution (two options here: from front of vehicle or from cog)
- ETA (derived?)
- 6 bits, crosstrack error +/- 150m
- ?? bits, airspeed error

**Comment by lvale** [comment]
> perhaps @floaledm and @yaapu could join us on the dev call? I see that it had the tag, and it was removed, but maybe it needs more discussion? @OXINARF you removed the tag, what do you think?
> 
> http://ardupilot.org/dev/docs/ardupilot-mumble-server.html

@magicrub , check the dev call notes

**Comment by floaledm** [comment]
@yaapu Thanks for the summary. Going back to my side note on scheduled tasks, I'm not clear on this: has the scheduled task method you're suggesting been recommended or cleared by Tridge or someone else (during the dev call or otherwise)? A couple of years ago, I suggested removing them because they were (and still are) unnecessary, and Tridge made code modifications to help me get that done. For example, this is what we did in Rover (same was done for Plane and Copter): https://github.com/ArduPilot/ardupilot/commit/40308512071bceccc6e5c713874465be6ad884b7#diff-d82d7c7fb685fcfe2e2d2b8b8fcb77bd

I personally don't see any benefit to scheduled tasks as it introduces unnecessary calculation cycles and the info is not as fresh as it could be... interested in hearing more about the "centralized" approach.

**Comment by OXINARF** [comment]
@magicrub I don't have anything to do add here. I removed the DevCall label because this was discussed in the dev call... If people want to discuss it again, label can be added again. In my opinion, deciding what to send is up for @floaledm to decide: the protocol was done by him and he's library maintainer.

**Comment by yaapu** [comment]
@floaledm proposal was to remove scheduled task in favor of new struct in AP_Vehicle.
There were no further details in the call, Peter Barker offered to write a prototype for the new struct and Tridge idea was to first port the OSD code to this new framework before porting other telemetry stacks (devo, frsky,etc..).
This PR has been put on hold until the new code (AP_Vehicle new struct + vehicle specific code) has been verified to work on OSD.
Testing on OSD is easier because it can be done 100% in sitl.

**Comment by magicrub** [comment]
@OXINARF I propose we make it a policy that when we remove the devcall label we add a comment with what the outcome was. Even if the outcome was to put it off, saying that is helpful to those that were not in the call.

**Comment by peterbarker** [comment]
@yaapu this looks like it will require a rather large amount of work to rebase.

Would this be better done as a reimplementation?  Is the data already provided in some other way?

Should we close this PR?

**Comment by yaapu** [comment]
@peterbarker the idea is to provide the same info via scripting:
- this adds the bindings for waypoint info https://github.com/ArduPilot/ardupilot/pull/13671
- this adds the binding for injecting telemetry data https://github.com/ArduPilot/ardupilot/pull/15416

Scripting would allow to experiment with multiple waypoints data structures without touching the code., so as far as I am concerned we can close this one

**Comment by peterbarker** [comment]
Thanks @yaapu .  Closing this one, and looking forward to more cool stuff :+1:
